### ISSUE TITLE
Updated inheritance test

### ISSUE THREAD

**Issue by mjanda** [djc/askama#240]
Hello,
I'm using deprecated `_parent` feature and trying to get rid of it but I'm not sure how.

I'm directly accessing fields on my parent template from both parent and child templates. If I change `_parent` to some other name, I get "no field" on either parent or child template.

And it looks like crate tests are still using this deprecated syntax, so no luck looking there. Is there any updated guide?

Thank you

**Comment by djc** [comment]
Is your code open source, or can you make a quick example? I'm not sure I fully understand how you're trying to fix it.

**Comment by mjanda** [comment]
Ok, I think I know where my problem is. I'm deriving Template for my parent templates too. I fixed that and mostly solved my problem, but still couldn't get it to work because my inheritance hierarchy sometimes uses two and sometimes three levels. Due to this, I can't reference anything in my top level template...because sometimes it would be `base.foo` and sometimes `base.base.foo`. With `_parent`, I can simply use `foo` and it will resolve to correct field.

Example here:
https://github.com/mjanda/askama-demo

**Comment by fahrradflucht** [comment]
Hey, I'm having trouble understanding the deprecation of `_parent` as well and thought maybe this is the place to ask. - What is the path forward from using `_parent`?

I'm currently using it like this (simplified example):
```rust
#[derive(Template)]
#[template(path = "layout/app.html")]
pub struct AppLayout<'a> {
  pub page_title: Option<&'a str>,
}

impl Default for AppLayout<'_> {
  fn default() -> Self {
    AppLayout {
      page_title: Default::default(),
    }
  }
}

#[derive(Template)]
#[template(path = "resource/new.html")]
pub struct NewTemplate<'a> {
    pub _parent: AppLayout<'a>,
}

fn render() -> Template {
    NewTemplate { _parent: Default::default() }
}
```

The idea was at the time (and still is?), that the layout provides sensible defaults and as long as you don't want to override something you just construct the layout `_parent` with `default()`.

So my question basically is. How do I achieve something with similar ergonomics without using the deprecated `_parent` field?

**Comment by djc** [comment]
Phew, I'm not sure if I can even remember correctly how it's different exactly. I think what happens is that in the new implementation, the inheritance is flattened at compile time, so that the resulting template acts as simple template with all the appropriate block contents from any parent (or grandparent, or deeper ancestry) templates is rendered; in turn the context type for the template should directly contain all the fields that these blocks rely upon. As such, in your example you'd just move the definition of `page_title` from `AppLayout` to `NewTemplate`, and remove the `_parent` definition from `NewTemplate`. You can remove all the source code for `AppLayout` (other than the template source), and it should just work.

**Comment by fahrradflucht** [comment]
Thanks for the reply. That's what I figured, but it is kind of sad, because I now have to duplicate all the default values from `AppLayout` onto every template.

**Comment by djc** [comment]
Nope! You can stuff them into a single type, a field of which you put in each of your template context types.

**Comment by fahrradflucht** [comment]
True! Thanks 🙏  I was mentally trapped in my existing setup 😉
